File: GoBackend/internal/controllers/ride_controller.go
package controllers

import (
	"uber/internal/models"
	"uber/internal/realtime"
	"uber/internal/services"

	"github.com/gin-gonic/gin"
	"go.mongodb.org/mongo-driver/bson/primitive"
)

// 💰 FARE ESTIMATE
func GetFare(c *gin.Context) {
	pickup := c.Query("pickup")
	destination := c.Query("destination")
	tripType := c.DefaultQuery("tripType", "one_way")
	if pickup == "" || destination == "" {
		c.JSON(400, gin.H{"error": "pickup and destination required"})
		return
	}

	fare, distanceKm, duration, _ := services.CalculateFare(pickup, destination, tripType)

	c.JSON(200, gin.H{
		"fare":     fare,
		"distance": distanceKm,
		"duration": duration,
	})
}

// 🚗 CREATE
func CreateRide(c *gin.Context) {

	var body struct {
		Pickup      string `json:"pickup"`
		Destination string `json:"destination"`
		VehicleType string `json:"vehicleType"`
		TripType    string `json:"tripType"`
	}

	if err := c.ShouldBindJSON(&body); err != nil {
		c.JSON(400, gin.H{"error": err.Error()})
		return
	}

	user := c.MustGet("user").(models.User)
	userID := user.ID

	tripType := body.TripType
	if tripType == "" {
		tripType = "one_way"
	}

	ride, err := services.CreateRide(userID, body.Pickup, body.Destination, body.VehicleType, tripType)
	if err != nil {
		c.JSON(500, gin.H{"error": err.Error()})
		return
	}

	if payload, err := services.BuildRidePayload(ride.ID); err == nil {
		realtime.BroadcastToCaptains("new-ride", payload)
	}

	c.JSON(201, ride)
}

// 🚗 CONFIRM
func ConfirmRide(c *gin.Context) {

	var body struct {
		RideID string `json:"rideId"`
	}

	if err := c.ShouldBindJSON(&body); err != nil {
		c.JSON(400, gin.H{"error": err.Error()})
		return
	}
	if body.RideID == "" {
		c.JSON(400, gin.H{"error": "rideId required"})
		return
	}

	captain := c.MustGet("user").(models.Captain)
	captainID := captain.ID

	rideID, err := primitive.ObjectIDFromHex(body.RideID)
	if err != nil {
		c.JSON(400, gin.H{"error": "invalid rideId"})
		return
	}

	ride, err := services.ConfirmRide(rideID, captainID)
	if err != nil {
		c.JSON(500, gin.H{"error": err.Error()})
		return
	}

	if payload, err := services.BuildRidePayload(ride.ID); err == nil {
		realtime.SendToUser(ride.User.Hex(), "ride-confirmed", payload)
	}

	c.JSON(200, ride)
}

// 🚗 START
func StartRide(c *gin.Context) {

	rideIDStr := c.Query("rideId")
	otp := c.Query("otp")
	if rideIDStr == "" || otp == "" {
		c.JSON(400, gin.H{"error": "rideId and otp required"})
		return
	}

	rideID, err := primitive.ObjectIDFromHex(rideIDStr)
	if err != nil {
		c.JSON(400, gin.H{"error": "invalid rideId"})
		return
	}

	ride, err := services.StartRide(rideID, otp)
	if err != nil {
		c.JSON(400, gin.H{"error": err.Error()})
		return
	}

	if payload, err := services.BuildRidePayload(ride.ID); err == nil {
		realtime.SendToUser(ride.User.Hex(), "ride-started", payload)
	}

	c.JSON(200, ride)
}

// 🏁 END
func EndRide(c *gin.Context) {

	var body struct {
		RideID string `json:"rideId"`
	}

	if err := c.ShouldBindJSON(&body); err != nil {
		c.JSON(400, gin.H{"error": err.Error()})
		return
	}
	if body.RideID == "" {
		c.JSON(400, gin.H{"error": "rideId required"})
		return
	}

	captain := c.MustGet("user").(models.Captain)
	captainID := captain.ID

	rideID, err := primitive.ObjectIDFromHex(body.RideID)
	if err != nil {
		c.JSON(400, gin.H{"error": "invalid rideId"})
		return
	}

	ride, err := services.EndRide(rideID, captainID)
	if err != nil {
		c.JSON(400, gin.H{"error": err.Error()})
		return
	}

	if payload, err := services.BuildRidePayload(ride.ID); err == nil {
		realtime.SendToUser(ride.User.Hex(), "ride-ended", payload)
	}

	c.JSON(200, ride)
}

// 💳 PAY
func PayRide(c *gin.Context) {
	var body struct {
		RideID string `json:"rideId"`
	}

	if err := c.ShouldBindJSON(&body); err != nil {
		c.JSON(400, gin.H{"error": err.Error()})
		return
	}
	if body.RideID == "" {
		c.JSON(400, gin.H{"error": "rideId required"})
		return
	}

	user := c.MustGet("user").(models.User)

	rideID, err := primitive.ObjectIDFromHex(body.RideID)
	if err != nil {
		c.JSON(400, gin.H{"error": "invalid rideId"})
		return
	}

	ride, err := services.MarkRidePaid(rideID, user.ID)
	if err != nil {
		c.JSON(400, gin.H{"error": err.Error()})
		return
	}

	if ride.Captain != nil {
		if payload, err := services.BuildRidePayload(ride.ID); err == nil {
			realtime.SendToCaptain(ride.Captain.Hex(), "ride-paid", payload)
		}
	}

	c.JSON(200, ride)
}

File: SpringServer/src/main/java/com/uber/springserver/controllers/RideController.java
package com.uber.springserver.controllers;

import com.uber.springserver.config.RequestContext;
import com.uber.springserver.dto.ride.CreateRideRequest;
import com.uber.springserver.dto.ride.RideActionRequest;
import com.uber.springserver.models.Captain;
import com.uber.springserver.models.Ride;
import com.uber.springserver.models.User;
import com.uber.springserver.services.RealtimeService;
import com.uber.springserver.services.RideService;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import jakarta.servlet.http.HttpServletRequest;
import java.util.Map;

@RestController
@RequestMapping("/api")
public class RideController {

    private final RideService rideService;
    private final RealtimeService realtimeService;

    public RideController(RideService rideService, RealtimeService realtimeService) {
        this.rideService = rideService;
        this.realtimeService = realtimeService;
    }

    @GetMapping({"/ride/get-fare", "/rides/get-fare"})
    public Object getFare(@RequestParam String pickup,
                          @RequestParam String destination,
                          @RequestParam(defaultValue = "one_way") String tripType) {
        if (pickup == null || pickup.isBlank() || destination == null || destination.isBlank()) {
            return Map.of("error", "pickup and destination required");
        }
        return rideService.calculateFare(pickup, destination, tripType);
    }

    @PostMapping({"/ride/create", "/rides/create"})
    public Object create(@RequestBody CreateRideRequest body, HttpServletRequest request) {
        User user = RequestContext.getUser(request);
        if (user == null) {
            return Map.of("message", "User not found");
        }

        String tripType = (body.getTripType() == null || body.getTripType().isBlank()) ? "one_way" : body.getTripType();

        try {
            Ride ride = rideService.createRide(user.getId(), body.getPickup(), body.getDestination(), body.getVehicleType(), tripType);
            try {
                realtimeService.broadcastToCaptains("new-ride", rideService.buildRidePayload(ride.getId()));
            } catch (Exception ignored) {
            }
            return ride;
        } catch (Exception e) {
            return Map.of("error", e.getMessage());
        }
    }

    @PostMapping({"/ride/confirm", "/rides/confirm"})
    public Object confirm(@RequestBody RideActionRequest body, HttpServletRequest request) {
        Captain captain = RequestContext.getCaptain(request);
        if (captain == null) {
            return Map.of("message", "Captain not found");
        }
        if (body.getRideId() == null || body.getRideId().isBlank()) {
            return Map.of("error", "rideId required");
        }

        try {
            Ride ride = rideService.confirmRide(body.getRideId(), captain.getId());
            try {
                realtimeService.sendToUser(ride.getUser(), "ride-confirmed", rideService.buildRidePayload(ride.getId()));
            } catch (Exception ignored) {
            }
            return ride;
        } catch (Exception e) {
            return Map.of("error", e.getMessage());
        }
    }

    @GetMapping({"/ride/start", "/rides/start-ride"})
    public Object start(@RequestParam String rideId, @RequestParam String otp) {
        if (rideId == null || rideId.isBlank() || otp == null || otp.isBlank()) {
            return Map.of("error", "rideId and otp required");
        }
        try {
            Ride ride = rideService.startRide(rideId, otp);
            try {
                realtimeService.sendToUser(ride.getUser(), "ride-started", rideService.buildRidePayload(ride.getId()));
            } catch (Exception ignored) {
            }
            return ride;
        } catch (Exception e) {
            return Map.of("error", e.getMessage());
        }
    }

    @PostMapping({"/ride/end", "/rides/end-ride"})
    public Object end(@RequestBody RideActionRequest body, HttpServletRequest request) {
        Captain captain = RequestContext.getCaptain(request);
        if (captain == null) {
            return Map.of("message", "Captain not found");
        }
        if (body.getRideId() == null || body.getRideId().isBlank()) {
            return Map.of("error", "rideId required");
        }

        try {
            Ride ride = rideService.endRide(body.getRideId(), captain.getId());
            try {
                realtimeService.sendToUser(ride.getUser(), "ride-ended", rideService.buildRidePayload(ride.getId()));
            } catch (Exception ignored) {
            }
            return ride;
        } catch (Exception e) {
            return Map.of("error", e.getMessage());
        }
    }

    @PostMapping({"/ride/pay", "/rides/pay"})
    public Object pay(@RequestBody RideActionRequest body, HttpServletRequest request) {
        User user = RequestContext.getUser(request);
        if (user == null) {
            return Map.of("message", "User not found");
        }
        if (body.getRideId() == null || body.getRideId().isBlank()) {
            return Map.of("error", "rideId required");
        }

        try {
            Ride ride = rideService.markRidePaid(body.getRideId(), user.getId());
            if (ride.getCaptain() != null && !ride.getCaptain().isBlank()) {
                try {
                    realtimeService.sendToCaptain(ride.getCaptain(), "ride-paid", rideService.buildRidePayload(ride.getId()));
                } catch (Exception ignored) {
                }
            }
            return ride;
        } catch (Exception e) {
            return Map.of("error", e.getMessage());
        }
    }
}
